paint-prototype/Assets/Fork/Tom/Scripts/PlayerMovement.cs

```diff
@@ -0,0 +1,281 @@
+    bool OnSteep => steepContactCount > 0;
+
+    int jumpPhase;
+
+    float minGroundDotProduct, minStairsDotProduct;
+    int stepsSinceLastGrounded, stepsSinceLastJump;
+
+
+    void OnValidate()
+    {
+        minGroundDotProduct = Mathf.Cos(maxGroundAngle * Mathf.Deg2Rad);
+        minStairsDotProduct = Mathf.Cos(maxStairsAngle * Mathf.Deg2Rad);
+    }
+
+    void Awake()
+    {
+        rBody = GetComponent<Rigidbody>();
+        rBody.freezeRotation = true;
+        OnValidate();
+    }
+
+    void Update()
+    {
+        PlayerInputs();
+        ControlDrag();
+
+    }
+
+    void FixedUpdate()
+    {
+        UpdateState();
+        AdjustVelocity();
+
+        if (desiredJump)
+        {
+            desiredJump = false;
+            Jump();
+        }
+
+        rBody.velocity = velocity;
+        ClearState();
+    }
+
+    void PlayerInputs()
+    {
+        hMovement = Input.GetAxisRaw("Horizontal");
+        vMovement = Input.GetAxisRaw("Vertical");
+
+        //moves on depending on the players facing
+        cameraDirection = transform.forward * vMovement + transform.right * hMovement;
+
+        desiredVelocity =
+            cameraDirection * maxSpeed;
+
+        desiredJump |= Input.GetButtonDown("Jump");
+    }
+
+    void ControlDrag()
+    {
+        rBody.drag = 0.5f;
+    }
+
+    void ClearState()
+    {
+        groundContactCount = steepContactCount = 0;
+        contactNormal = steepNormal = Vector3.zero;
+    }
+
+    void UpdateState()
+    {
+        stepsSinceLastGrounded += 1;
+        stepsSinceLastJump += 1;
+        velocity = rBody.velocity;
+        if (OnGround || SnapToGround() || CheckSteepContacts())
+        {
+            stepsSinceLastGrounded = 0;
+            if (stepsSinceLastJump > 1)
+            {
+                jumpPhase = 0;
+            }
+            if (groundContactCount > 1)
+            {
+                contactNormal.Normalize();
+            }
+        }
+        else
+        {
+            contactNormal = Vector3.up;
+        }
+    }
+
+    bool SnapToGround()
+    {
+        if (stepsSinceLastGrounded > 1 || stepsSinceLastJump <= 2)
+        {
+            return false;
+        }
+        float speed = velocity.magnitude;
+        if (speed > maxSnapSpeed)
+        {
+            return false;
+        }
+        if (!Physics.Raycast(
+            rBody.position, Vector3.down, out RaycastHit hit,
+            probeDistance, probeMask
+        ))
+        {
+            return false;
+        }
+        if (hit.normal.y < GetMinDot(hit.collider.gameObject.layer))
+        {
+            return false;
+        }
+
+        groundContactCount = 1;
+        contactNormal = hit.normal;
+        float dot = Vector3.Dot(velocity, hit.normal);
+        if (dot > 0f)
+        {
+            velocity = (velocity - hit.normal * dot).normalized * speed;
+        }
+        return true;
+    }
+
+    bool CheckSteepContacts()
+    {
+        if (steepContactCount > 1)
```

Comment:
To avoid doing a nested if, we can early return false, since we are not doing anything special in the `else` case.